tests/toranj/test-020-router-table.py

```diff
@@ -0,0 +1,198 @@
+# Init all nodes
+
+wpan.Node.init_all_nodes()
+
+#-----------------------------------------------------------------------------------------------------------------------
+# Build network topology
+#
+#
+#     r1 ------ r2
+#      \        /
+#       \      /
+#        \    /
+#          r3 _____ r4
+#
+#
+
+r1.form("route-table")
+
+r1.whitelist_node(r2)
+r2.whitelist_node(r1)
+r2.join_node(r1, wpan.JOIN_TYPE_ROUTER)
+
+r2.whitelist_node(r3)
+r3.whitelist_node(r2)
+r3.join_node(r2, wpan.JOIN_TYPE_ROUTER)
+
+r3.whitelist_node(r1)
+r1.whitelist_node(r3)
+
+r3.whitelist_node(r4)
+r4.whitelist_node(r3)
+r4.join_node(r3, wpan.JOIN_TYPE_ROUTER)
+
+# c4 is attached to r4 so that it quickly gets promoted to a router role.
+c4.whitelist_node(r4)
+r4.whitelist_node(c4)
+c4.join_node(r4, wpan.JOIN_TYPE_SLEEPY_END_DEVICE)
+c4.set(wpan.WPAN_POLL_INTERVAL, '2000')
+
+#-----------------------------------------------------------------------------------------------------------------------
+# Test implementation
+#
+
+verify(r1.get(wpan.WPAN_NODE_TYPE) == wpan.NODE_TYPE_LEADER)
+verify(r2.get(wpan.WPAN_NODE_TYPE) == wpan.NODE_TYPE_ROUTER)
+verify(r3.get(wpan.WPAN_NODE_TYPE) == wpan.NODE_TYPE_ROUTER)
+verify(r4.get(wpan.WPAN_NODE_TYPE) == wpan.NODE_TYPE_ROUTER)
+
+r1_id = int(r1.get(wpan.WPAN_THREAD_ROUTER_ID), 0)
+r2_id = int(r2.get(wpan.WPAN_THREAD_ROUTER_ID), 0)
+r3_id = int(r3.get(wpan.WPAN_THREAD_ROUTER_ID), 0)
+r4_id = int(r4.get(wpan.WPAN_THREAD_ROUTER_ID), 0)
+
+r1_ext_addr = r1.get(wpan.WPAN_EXT_ADDRESS)[1:-1]
+r2_ext_addr = r2.get(wpan.WPAN_EXT_ADDRESS)[1:-1]
+r3_ext_addr = r3.get(wpan.WPAN_EXT_ADDRESS)[1:-1]
+r4_ext_addr = r4.get(wpan.WPAN_EXT_ADDRESS)[1:-1]
+
+r1_rloc = int(r1.get(wpan.WPAN_THREAD_RLOC16), 16)
+r2_rloc = int(r2.get(wpan.WPAN_THREAD_RLOC16), 16)
+r3_rloc = int(r3.get(wpan.WPAN_THREAD_RLOC16), 16)
+r4_rloc = int(r4.get(wpan.WPAN_THREAD_RLOC16), 16)
+
+WAIT_TIME = 30 / speedup + 5
+
+INVALID_ROUTER_ID = 63
+
+def check_r1_router_table():
+    router_table = wpan.parse_router_table_result(r1.get(wpan.WPAN_THREAD_ROUTER_TABLE))
+    verify(len(router_table) == 4)
+    for entry in router_table:
+        if entry.rloc16 == r1_rloc:
+            pass
+        elif entry.rloc16 == r2_rloc:
+            # r1 should be directly connected to r2.
+            verify(entry.is_link_established())
+            verify(entry.ext_address == r2_ext_addr)
+        elif entry.rloc16 == r3_rloc:
+            # r1 should be directly connected to r3.
+            verify(entry.is_link_established())
+            verify(entry.ext_address == r3_ext_addr)
+        elif entry.rloc16 == r4_rloc:
+            # r1's next hop towards r4 should be through r3.
+            verify(not entry.is_link_established());
+            verify(entry.next_hop == r3_id)
+        else:
+            raise(wpan.VerifyError("unknown entry in the router table of r1"))
+
+wpan.verify_within(check_r1_router_table, WAIT_TIME)
```

Comment:
I don't see a validation of R2's router table. Was that intentional given that it's pretty much the same as R1?

Reply:
Exactly. It's a mirror of R1...